rust/actyx/ax-aql/src/language/parser.rs

```diff
@@ -721,6 +722,116 @@ pub(crate) fn query_from_str(s: &str) -> Result<Query<'_>> {
     Ok(query)
 }
 
+pub(crate) struct QueryWorkflowAnalysisError<'a> {
+    pub(crate) span: Span<'a, Ident>,
+    pub(crate) error_type: QueryWorkflowAnalysisErrorType,
+}
+
+pub(crate) enum QueryWorkflowAnalysisErrorType {
+    UndeclaredEvent,
+    UndeclaredWorkflowCall,
+}
+
+pub(crate) struct QueryWorkflowAnalysis<'a> {
+    pub(crate) query: &'a Query<'a>,
+    pub(crate) workflow_tracker: WorkflowTracker<'a>,
+}
+
+impl<'a> QueryWorkflowAnalysis<'a> {
+    pub fn from(query: &'a Query) -> Self {
+        Self {
+            query,
+            workflow_tracker: Default::default(),
+        }
+    }
+
+    pub fn check(&'a self, query: &'a Query) -> Vec<QueryWorkflowAnalysisError> {
```

Comment:
The `query` argument should be the same as `self.query`, right? In that case it should be removed.

Reply:
Ah yes.